twapake.man
.TH "twapake" 1 "06/30/2012" "Frederic Galusik"


.SH NAME

.P
twapake \- A simple perl tool to manage last installed packages and to
keep a log of packages installed date

.SH SYNOPSIS

.P
\fBtwapake\fR [\fIOPTIONS\fR]

.SH DESCRIPTION

.P
\fBTwapake\fR is a system administration utility for Slackware
and Slackware based operating systems only, making easier the management
of last installed packages and making easier seeing the installation
date of a package.

.SH GENERAL OPTIONS

.TP
\fB\-h\fR, \fB\-\-help\fR
Simple help

.TP
\fB\-v\fR, \fB\-\-version\fR
Print Twapake version

.TP
\fB\-l\fR, \fB\-\-logfile\fR
Print a snapshot logfile to the /var/log/twapake/ directory.
A snapshot is a list of all packages sorted by installed date at T time.
The installed date is printed too, making easier to see when a package
has been installed on your system.

.TP
\fB\-i\fR, \fB\-\-installed\fR
Print a simple sorted list of installed packages to STDOUT

.TP
\fB\-n\fR, \fB\-\-nlpkg\fR
Used with \fB\-i\fR. Let you choose how many \fBN\fR last installed packages
you want to print to STDOUT

.TP
\fB\-r\fR, \fB\-\-remove\fR
Remove the last \fBN\fR installed packages

.TP
\fB\-s\fR, \fB\-\-since\fR
Print today installed packages to STDOUT

.TP
\fB\-c\fR, \fB\-\-choosedate\fR
Used with \fB\-s\fR. Print packages installed since this choosen date

.TP
\fB\-u\fR, \fB\-\-ui\fR
Launch the Curses User Interface for an easier removal. You can choose
the packages you want to remove in a list sorted by installed date

.SH COMMAND LINE USAGE

.P
Here are some usages examples:

.TP
\fBtwapake \-i \-n 10\fR
Print the 10 last installed packages to STDOUT

.TP
\fBtwapake \-r \-n 5\fR
Show you the last 5 installed packages and ask you if you want to remove
them

.TP
\fBtwapake \-s \-c 2011\-05\-25\fR
Show you the packages installed on your system since the 2011\-05\-25

.SH THE CURSES INTERFACE

.P
Within the program you can use the following keys.

.TP
\fBUP / DOWN\fR
Navigate the packages list

.TP
\fBSPACE\fR
Select/Unselect a package

.TP
\fBCtrl\-X\fR
Show you the menu

.TP
\fBCtrl\-U\fR
Unselect all choosen packages

.TP
\fBCtrl\-R\fR
Remove all selected packages

.TP
\fBCtrl\-Q\fR
Quit Twapake

.SH WEBSITE

.P
http://www.salixos.org/wiki/index.php/Twapake

.SH AUTHOR

.P
Twapake was written by Frederic Galusik <fredg~at~salixos~dot~org>

.SH SEE ALSO

.P
\fBpkgtool\fR(8), \fBremovepkg\fR(8), \fBspkg\fR(8)

.\" man code generated by txt2tags 2.6 (http://txt2tags.org)
.\" cmdline: txt2tags twapake.t2t


ChangeLog

* version 0.5:
    - add --ui (-u): launch the Curses User Interface for an easy removal
    - add a man page
    - made Twapake and Twacurse perl modules

* version 0.4:
    - add --since (-s): print today installed packages
    - add -c to use with -s: print installed packages since a choosen date
    - change the date format of the twalog file to be more internationnal

* version 0.3:
    - add --remove (-r) to use with -n : remove the last N installed
    packages.
    - use spkg by default else removepkg for removing packages
    - add cron job

* version 0.2:
    - add a --nlpkg (-n) option to use with -i to specify how many
    last packages have to be printed to STDOUT
    - change the output format of the twapake snapshot to be more
    human readable

* version 0.1.0.1:
    - some typos

* version 0.1:
    - first try
    - --logfile (-l) print a twapake snapshot logfile
    - --installed (-i) print a simple sorted list to STDOUT


Twapake.pm
package Twapake;

#       Twapake.pm
#
#       This file is part of twapake
#
#       Copyright 2010-2011 Frédéric Galusik <fredg~at~salixos~dot~org>
#       License: BSD Revised
#
###############################################################################

use strict;
use warnings;
use POSIX;
use File::stat;
use File::Path;

#
# The place where Slackware store installed packages log files
#
our $pkgdir = "/var/log/packages";

our $rm_command;
our $removepkg = "/sbin/removepkg";
our $spkg = "/sbin/spkg";
our $spkgd = "/sbin/spkg -d";

our $file;
our @files;
our @sortfiles;
our ( $a, $b );
our $now = localtime;
our $nowfile = strftime("%Y-%m-%d-%H-%M-%S", localtime(time));

# the place to store twapake log files
our $twa_path = "/var/log/twapake";
mkpath $twa_path, 1, 0755;

sub new {
    my $self = {};
    bless ($self);
    return $self;
}

#
# check if we use spkg or removepkg
#
sub checkrm {
    if ( -e $spkg and -e $removepkg ) {
        $rm_command = $spkgd;
        } elsif (-e $removepkg and not -e $spkg ) {
            $rm_command = $removepkg;
            } else {
                die "Nor pkgtool or spkg are installed, aborting...\n";
            }
}

#
# get the packages list
#
sub get_files {
    opendir(DIR, $pkgdir) or die "Cannot open '$pkgdir' : $!\n";
    @files = grep !/^\./, readdir(DIR);
    closedir(DIR);
    return @files;
}
get_files();

#
# get the total of installed packages
#
sub nbpkg {
    our $npkg = @files;
    return $npkg;
}

#
# build a hash with key = pkg name and value = last time modification
#
keys our %hache = @files;
foreach $file (@files) {
    our $stpkg = stat("$pkgdir/$file");
    $hache{$file} = $stpkg->mtime;
}

#
# sort hash by descending value date
#
sub sortfiles {
    @sortfiles = sort { $hache{$b} <=> $hache{$a} } @files;
    return @sortfiles;
}
sortfiles();

#
# make the twapake snapshot:
# installed date - package name
#
sub twalog {
    open(TWAFILE, ">$twa_path/${nowfile}_twapake.log");
    print TWAFILE "\nTwapake snapshot on $now\n";
    print TWAFILE '='x72 ."\n\n";
    print TWAFILE "Installed dates\t\tPackages (total: ", nbpkg(), ")\n";
    print TWAFILE '-'x19 ."\t" . '-'x45 ."\n";
    for $file (@sortfiles) {
        printf TWAFILE "%s\t%s\n",
        strftime("%Y-%m-%d %H:%M:%S", localtime($hache{$file})), $file;
    }
    close(TWAFILE);
}

#
# make a less detailed list
# give the choice of number of packages
#
our $nlpkg;
$nlpkg = nbpkg() unless defined($nlpkg);
sub checknpkg {
    if ($nlpkg > nbpkg()) {
        die "Error: only ", nbpkg(), " packages are installed.\n";
    }
}

sub sortmtimelist {
    checknpkg();
    foreach $file ( 0..($nlpkg - 1) ) {
        print "$sortfiles[$file]\n";
    }
}

#
# remove last installed packages
#
sub rmpkg {
    checknpkg();
    checkrm();
    print "Do you wish to remove all these packages ?\n";
    foreach $file ( 0..($nlpkg - 1) ) {
        print "$sortfiles[$file]\n";
    }
    print "(y)es or (n)o ? ";
    if ((my $answer = <STDIN>) =~ /^y$/i) {
        foreach $file ( 0..($nlpkg - 1) ) {
            system ("$rm_command $sortfiles[$file]");
        }
    } elsif ($answer =~ /^n$/i) {
        die "aborting ...\n";
    } else {
        chomp $answer;
        die "'$answer' is neither 'y' for 'yes' nor 'n' for 'no'\n";
    }
}

#
# show installed packages since a choosen date
#
our $choosedate;
sub userdate_default {
    my $today = strftime("%Y-%m-%d", localtime(time));
    return $today;
}
$choosedate = userdate_default() unless defined($choosedate);
sub checkdate {
    if ($choosedate =~ m!^((?:19|20)\d\d)[- /.](0[1-9]|1[012])[- /.](0[1-9]|[12][0-9]|3[01])$!) {
    # $1 year $2 month $3 day
    if ($3 == 31 and ($2 == 4 or $2 == 6 or $2 == 9 or $2 == 11)) {
      die "Sorry but this month has not 31 days!\n";
    } elsif ($3 >= 30 and $2 == 2) {
      die "Sorry but February has not as much days!\n";
    } elsif ($2 == 2 and $3 == 29 and not ($1 % 4 == 0 and ($1 % 100 != 0 or $1 % 400 == 0))) {
      die "Sorry but it is not a leap year\n";
    }
  } else {
    die "Sorry but date format must be YYYY-mm-dd!\n";
  }
}

sub installedsince {
    for $file (@sortfiles) {
        if (strftime("%Y-%m-%d %H:%M:%S", localtime($hache{$file})) gt $choosedate) {
            print "$file\n";
        }
    }
}

1;
